test/parse5_utils_test.js

```diff
@@ -69,4 +69,154 @@ suite('parse5_utils', function() {
+      fn = parse5_utils.predicates.hasClass('d');
+      assert.isFalse(fn(frag));
+    });
+
+    test('AND', function() {
+      var preds = [
+        parse5_utils.predicates.hasTagName('div'),
+        parse5_utils.predicates.hasAttrValue('id', 'a'),
+        parse5_utils.predicates.hasClass('b')
+      ];
+      var fn = parse5_utils.predicates.AND(preds);
+      assert.isFunction(fn);
+      assert.isTrue(fn(frag));
+      preds.push(parse5_utils.predicates.hasClass('d'));
+      assert.isFalse(fn(frag));
+    });
+
+    test('OR', function() {
+       var preds = [
+        parse5_utils.predicates.hasTagName('div'),
+        parse5_utils.predicates.hasAttr('hidden')
+      ];
+      var fn = parse5_utils.predicates.OR(preds);
+      assert.isFunction(fn);
+      assert.isTrue(fn(frag));
+      preds.shift();
+      fn = parse5_utils.predicates.OR(preds);
+      assert.isFalse(fn(frag));
+   });
+
+   test('NOT', function() {
+     var pred = parse5_utils.predicates.hasTagName('a');
+     var fn = parse5_utils.predicates.NOT(pred);
+     assert.isFunction(fn);
+     assert.isTrue(fn(frag));
+     assert.isFalse(pred(frag));
+   });
+
+   test('Chaining Predicates', function() {
+     var fn = parse5_utils.predicates.AND([
+       parse5_utils.predicates.hasTagName('div'),
+       parse5_utils.predicates.OR([
+         parse5_utils.predicates.hasClass('b'),
+         parse5_utils.predicates.hasClass('d'),
+       ]),
+       parse5_utils.predicates.NOT(
+         parse5_utils.predicates.hasAttr('hidden')
+       )
+     ]);
+
+     assert.isFunction(fn);
+     assert.isTrue(fn(frag));
+   });
+  });
+
+  suite('Query', function() {
+    var docText = [
+      '<!DOCTYPE html>',
+      '<link rel="import" href="polymer.html">',
+      '<dom-module id="my-el">',
+      '<template>',
+      '<img src="foo.jpg">',
+      '<a href="next-page.html">',
+      '</template>',
+      '</dom-module>',
+      '<script>Polymer({is: "my-el"})</script>'
+    ].join('\n');
+    var doc = null;
+
+    setup(function() {
+      doc = parser.parse(docText);
+    });
+
+    test('query', function() {
+      var fn = parse5_utils.predicates.AND([
+        parse5_utils.predicates.hasTagName('link'),
+        parse5_utils.predicates.hasAttrValue('rel', 'import'),
+        parse5_utils.predicates.hasAttr('href')
+      ]);
+      var expected = doc.childNodes[1].childNodes[0].childNodes[0];
+      var actual = parse5_utils.query(doc, fn);
+      assert.equal(expected, actual);
+    });
+
+    test('queryAll', function() {
+      var fn = parse5_utils.predicates.AND([
+        parse5_utils.predicates.OR([
+          parse5_utils.predicates.hasAttr('href'),
+          parse5_utils.predicates.hasAttr('src'),
+        ]),
+        parse5_utils.predicates.NOT(
+          parse5_utils.predicates.hasTagName('link')
+        )
+      ]);
+
+      var templateContent = doc.childNodes[1].childNodes[1].childNodes[0].childNodes[1].childNodes[0];
```

Comment:
I've been adding comments to `childNodes` chains like this to say which nodes we're expecting so that it's easier to update later.